### PR TITLE
task/error-handling

### PR DESCRIPTION
**Pull request by makeitvsolo**
Added:
- Result API
- Optional API

### CODE REVIEW

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            _value = new TestValue();
+            _optional = Optional<TestValue>.Of(_value);
+        }
+
+        [Fact]
+        public void IsPresent_ReturnsTrue()
+        {
+            Assert.True(_optional.IsPresent);
+        }
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:23]: it's best to stick to the one-line style with => + use FluentAssertions with syntax example
`_optional.IsPresent.Should().Be(true)`
Also use different naming convention for test methods -> [click](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices)
```
The name of your test should consist of three parts:

    The name of the method being tested.
    The scenario under which it's being tested.
    The expected behavior when the scenario is invoked.
```

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
+using TaskNotFound.Shared.Abstract.ErrorHandling.Exceptions;
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:1]: Namespace does not correspond to file location, must be: `'Shared.Abstract.UnitTests.ErrorHandling.Optionals'` by ReSharper.
Also this problem in another files.

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+using Xunit;
+
+namespace TaskNotFound.Shared.Abstract.ErrorHandling.Optionals;
+
+public class OptionalTests
+{
+    public class WhenOptionalIsPresent
+    {
+        private TestValue _value;
+        private Optional<TestValue> _optional;
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:11]: Fields can be made readonly

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+        public WhenOptionalIsPresent()
+        {
+            _value = new TestValue();
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:15]: Can be simply with `_value = new();`

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            var mappedOptional = _optional.Or(() => optional);
+
+            Assert.NotEqual(optional, mappedOptional);
+        }
+
+        [Fact]
+        public void WhenPresent_DoingAction()
+        {
+            var count = 0;
+            var expected = 1;
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:98]: Convert this variable to constant (example: `const int expected = 0;`)

File: Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+using Xunit;
+
+namespace TaskNotFound.Shared.Abstract.ErrorHandling.Results;
+
+public class ResultTests
+{
+    public class WhenResultIsOk
+    {
+        private TestValue _value;
+        private Result<TestValue, TestError> _result;
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs:11]: Fields can be readonly.

File: Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+        public WhenResultIsOk()
+        {
+            _value = new TestValue();
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs:15]: Can be simply with `_value = new();`

File: Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            _result.WhenError(_ => count += 1);
+
+            Assert.Equal(expected, count);
+        }
+    }
+
+    public class WhenResultIsError
+    {
+        private TestError _error;
+        private Result<TestValue, TestError> _result;
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs:137]: Fields can be readonly.

File: Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+        public WhenResultIsError()
+        {
+            _error = new TestError();
```
Comment (shishnk) [inline Shared.Abstract.UnitTests/ErrorHandling/Results/ResultTests.cs:141]: Can be simply with `_error = new();`

File: Shared.Abstract/ErrorHandling/Optionals/Empty.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
+using TaskNotFound.Shared.Abstract.ErrorHandling.Exceptions;
+
+namespace TaskNotFound.Shared.Abstract.ErrorHandling.Optionals;
+
+internal sealed class Empty<T> : Optional<T>
+{
+    internal Empty()
+    {
+    }
```
Comment (shishnk) [inline Shared.Abstract/ErrorHandling/Optionals/Empty.cs:9]: Empty constructor is redundant. The compiler generates the same by default.

**Review by shishnk — CHANGES_REQUESTED** [review]

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs (hunk 2 above)
Comment (makeitvsolo) [reply Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:1]: Wtf is `Shared.Abstract`?
This library belongs to the project `TaskNotFound`.

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs (hunk 3 above)
Comment (makeitvsolo) [reply Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:11]: I wouldn't follow the rules in test projects.

File: Shared.Abstract/ErrorHandling/Optionals/Empty.cs (hunk 10 above)
Comment (makeitvsolo) [reply Shared.Abstract/ErrorHandling/Optionals/Empty.cs:9]: Default construct is public, but i want internal constructor

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs (hunk 3 above)
Comment (makeitvsolo) [reply Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:11]: > I wouldn't follow the rules in test projects.

I found out that I don't write in java, so no problem making them readonly

File: Shared.Abstract/ErrorHandling/Optionals/Empty.cs (hunk 10 above)
Comment (shishnk) [reply Shared.Abstract/ErrorHandling/Optionals/Empty.cs:9]: Is truth.

File: Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs (hunk 2 above)
Comment (shishnk) [reply Shared.Abstract.UnitTests/ErrorHandling/Optionals/OptionalTests.cs:1]: Will be fixed by ignoring the warning.

**Review by shishnk — APPROVED** [review]

**Pull request merged**